bidict/_common.py

```diff
@@ -53,103 +102,164 @@ def __inverted__(self):
+            if key_clbhv is RAISE:
+                # since multiple values can have the same hash value, refer
+                # to the existing key via `_inv[oldval]` rather than `key`
+                raise KeyExistsError((_inv[oldval], oldval))
+            elif key_clbhv is IGNORE:
+                return
         valexists = oldkey is not _missing
-        if valexists and not overwrite_key:
-            # since multiple values can have the same hash value,
-            # refer to the existing value via self._fwd[oldkey]
-            raise ValueExistsException((oldkey, self._fwd[oldkey]))
-        if keyexists:  # overwrite_val == True
-            del self._inv[oldval]
-        if valexists:  # overwrite_key == True
-            del self._fwd[oldkey]
-        self._fwd[key] = val
-        self._inv[val] = key
-
-    def _update(self, *args, **kw):
-        for k, v in pairs(*args, **kw):
-            self._put(k, v, overwrite_key=False, overwrite_val=True)
-
-    get = lambda self, k, *args: self._fwd.get(k, *args)
-    copy = lambda self: self.__class__(self._fwd)
-    get.__doc__ = 'Like :py:meth:`dict.get`.'
-    copy.__doc__ = 'Like :py:meth:`dict.copy`.'
-    __len__ = lambda self: len(self._fwd)
-    __iter__ = lambda self: iter(self._fwd)
-    __contains__ = lambda self, x: x in self._fwd
-    __len__.__doc__ = 'Like :py:meth:`dict.__len__`.'
-    __iter__.__doc__ = 'Like :py:meth:`dict.__iter__`.'
-    __contains__.__doc__ = 'Like :py:meth:`dict.__contains__`.'
-    keys = lambda self: self._fwd.keys()
-    items = lambda self: self._fwd.items()
-    keys.__doc__ = 'Like :py:meth:`dict.keys`.'
-    items.__doc__ = 'Like :py:meth:`dict.items`.'
-    values = lambda self: self._inv.keys()
+        if valexists:
+            if val_clbhv is RAISE:
+                # since multiple values can have the same hash value, refer
+                # to the existing value via `_fwd[oldkey]` rather than `val`
+                raise ValueExistsError((oldkey, _fwd[oldkey]))
+            elif val_clbhv is IGNORE:
+                return
+        _fwd.pop(oldkey, None)
+        _inv.pop(oldval, None)
+        _fwd[key] = val
+        _inv[val] = key
+
+    def _update(self, key_clbhv, val_clbhv, *args, **kw):
+        if not args and not kw:
+            return
+
+        _fwd, _inv, _missing = self._fwd, self._inv, self._missing
+
+        # take fast path if passed only another bidict
+        if not kw and args and isinstance(args[0], BidirectionalMapping):
+            l = len(args)
+            if l != 1:
+                raise TypeError('Expected at most 1 argument, got %d' % l)
+            updatefwd = self._dcls(args[0]._fwd)
+            updateinv = self._dcls(args[0]._inv)
+        else:
+            if key_clbhv is RAISE:
+                items = frozenset(pairs(*args, **kw))
+                updatefwd = self._dcls(items)
+                if len(items) > len(updatefwd):
+                    raise NonuniqueKeysError(items)
```

Comment:
The capitalisation of this exception class is strange. "nonunique" is not a word, is it? Could be better to spell it unambiguously like KeysNotUniqueException which also puts the important word "Keys" first.